### ISSUE TITLE
How to update firefly iii docker instance without losing transactions/history

### ISSUE THREAD

**Issue by 8bitgentleman** [firefly-iii/firefly-iii#999]
I run Firefly III from a docker container and I just realized that I am running an old version of Firefly III, version 4.6.2. I'd like to update to the latest version but I'm not sure how to do that without losing my transaction history. When I set up the docker instance with the directions from [https://firefly-iii.github.io/using-docker.html](https://firefly-iii.github.io/using-docker.html) I ended up with two separate containers and I don't fully understand how updating will affect them. I assume the DB container has all my data and the APP container is the one that should be updated? Is that right? If so how do I go about doing that?
<img width="291" alt="screen shot 2017-11-17 at 10 43 58 am" src="https://user-images.githubusercontent.com/4028391/32955661-a53ddc70-cb84-11e7-9d24-9d92b42af1ef.png">
 

Sorry if this has been answered before, I did a quick look through the closed issues and didn't find a clear answer. I've really enjoyed using firefly iii and I'd like to keep up to date with the awesome work you put into it :)

**Comment by JC5** [comment]
I have to admit I have no idea. I don't use docker myself. 

Allow me to ping some users who have helped me with Docker before.

@alazare619 @Jumanjii @TomWis97 @Moltenkore 

I hope they have the time to point you in the right direction. I shall update the website to include their instructions 👍

**Comment by alazare619** [comment]
Which direction section did you use. If you used the docker-compose method the app and db are seperate..so if you git cloned the repo then did docker-compose up app and db you can just git pull the diffs
the db will not change your db schema and everything will stay the same as its residing in a seperate mysql8 container called firefly-db_1 that wont change the repo only contains updates for just firefly-app

tl;dr git pull then docker-compose up fireflyapp fireflydb or whatever method you previously you did another method please respond so I can better assist you :)

**Comment by TomWis97** [comment]
Interestingly, the `docker-compose.prod.yml` only defines one container.

Some basics of container: A container should be stateless: never, ever, should data be saved to the docker container itself. If you need to store data, write it to the volume. When a new version of the container is released, just delete the old container, pull the latest version of the image and create a new container. The new container should point to the same volume as the previous version.

Settings like MySQL host, db, username and password should be passed through to the container as environment variables.

Some of the Docker-compose files define a volume at `/var/www/firefly-iii/storage`, but this volume isn't defined in `Dockerfile`. If firefly-iii really stores data in that directory, it should be defined as a volume while building the container. ( @JC5 should be able to answer this.)

Basically, updating to a new container version boils down to:
1. Take notes of the current environment variables and volumes of the running Docker container. (But you should have this somewhere already, IMHO.)
1. Stop and remove the old Docker container.
1. Pull the latest version of the Docker image.
1. Launch a new container with the same environment variables and volumes.

You should see the MySQL container as a separate entity. If a new version of the MySQL container is released, update the MySQL container. If a new version of the Firefly-iii container is released, update that container.

These steps might be somewhat different when using Docker compose, as I don't use docker-compose myself. (Because I already have a MySQL VM running for multiple applications.) But I don't expect huge changes.

I haven't had the time to look into the Docker container for Firefly-iii lately, as I'm busy with college stuff.

**Comment by alazare619** [comment]
@TomWis97  https://github.com/firefly-iii/firefly-iii/blob/master/docker-compose.yml

**Comment by 8bitgentleman** [comment]
@alazare619 It was a while ago that I set this up but I believe that I used the "For Production" commands with docker-compose. Just to confirm: I'll shut down my current running version of firefly iii and then start with the commands you suggested. You said do `git pull https://github.com/firefly-iii/firefly-iii.git`, should I do that or `docker pull jc5x/firefly-iii` like the [setup docs](https://firefly-iii.github.io/using-docker.html)?

After that won't doing `docker-compose up fireflyapp fireflydb` overwrite both the app and the db?

**Comment by alazare619** [comment]
Paste an LS of the directory your docker-compose files are located and gist them or code qoute all of hte docker compose files of the directory. This would give me a better idea of what we are working with. It seems to me that more then likely this is a very very old installation of docker and how it was setup before fireflyiii was integrated iwth dockerhub.

**Comment by 8bitgentleman** [comment]
[@alazare619 Here is an LS of /var/www/firefly-iii inside the firefly-app docker container.](https://gist.github.com/8bitgentleman/1b6c69025d5c093611ce18678bb6d460) Is that what you meant?

**Comment by alazare619** [comment]
alright do a cat docker-compose.yml and docker-compose.override.yml and docker-compose.prod.yml remove any private info such as passwords and such

**Comment by 8bitgentleman** [comment]
@alazare619 ok I ran those commands, do you want the output of one/all of them?

**Comment by alazare619** [comment]
all of them but as i said scrub the password info and stuff

**Comment by 8bitgentleman** [comment]
There didn't seem to be any person info from those commands so here's a gist of the output

https://gist.github.com/8bitgentleman/bfdef4cdcbda029ab77d22e7959df73c

**Comment by alazare619** [comment]
Alright so you are perfectly fine to upgrade heres what you need to do. First remove all files other then docker-compose.yml

Then replace the contents of docker compose with this.

`version: '2'
services:
  firefly-db:
    image: mysql:8
    environment:
      - MYSQL_DATABASE=firefly_db
      - MYSQL_USER=firefly_db
      - MYSQL_PASSWORD=firefly_db_secret
      - MYSQL_RANDOM_ROOT_PASSWORD=yes
    volumes:
      - firefly-storage:/var/lib/mysql

  firefly-app:
    image: 'jc5x/firefly-iii'
    environment:
      - FF_DB_HOST=firefly-db
      - FF_DB_NAME=firefly_db
      - FF_DB_USER=firefly_db
      - FF_DB_PASSWORD=firefly_db_secret
      - FF_APP_KEY=SomeRandomStringOf32CharsExactly
      - FF_APP_ENV=local
    ports:
      - "80:80"
    links:
      - firefly-db
    volumes:
      - firefly-app-storage:/var/www/firefly-iii/storage
volumes:
  firefly-storage:
    driver: local
  firefly-app-storage:
    driver: local`

**Comment by 8bitgentleman** [comment]
Ok so I screwed something up and the container stopped. I can't run any of those commands because when I try to start the firefly-app container again it says "Container does not exist." My data should be OK since the firefly-db container is untouched but what should I do to get the app running again?

**Comment by alazare619** [comment]
docker rm firefly-app then docker-compose up

**Comment by 8bitgentleman** [comment]
I ran those two commands and it created a new app container and a new db container. When I navigate to the port that firefly is exposed on it just says "Be right back" ?

**Comment by alazare619** [comment]
strange when your machine would reboot how did you start back up the containers?

**Comment by 8bitgentleman** [comment]
The containers started automatically after running docker-compose up. After rebooting them I get the same message, I used the graphical interface to turn them on and off if that makes a difference

**Comment by 8bitgentleman** [comment]
@JC5 and insight into the "Be right back" error? I tried the suggestions at #889 but that didn't change anything

**Comment by JC5** [comment]
There should be a notification in your logs what the error is. Laravel is the framework that throws the error and it's used on many occasions. Checkout `storage/logs` or the logs generated by Apache.

**Comment by 8bitgentleman** [comment]
It's saying something about an "unknown database firefly_db" even though the firefly_db container is still running
[ff3-apache2handler-2017-11-18.log](https://github.com/firefly-iii/firefly-iii/files/1484911/ff3-apache2handler-2017-11-18.log)

**Comment by JC5** [comment]
If the containers were re-created then there is no database anymore: it is lost.

**Comment by 8bitgentleman** [comment]
A new db folder was created when I used the `docker-container up` but the old db container is still there. Maybe I can link the old db container to the new app container?
<img width="391" alt="screen shot 2017-11-18 at 1 43 26 pm" src="https://user-images.githubusercontent.com/4028391/32983690-a6c481ec-cc66-11e7-9943-a86421bffdc9.png">

Regardless, when I navigate to the new firefly instance all the page says is "Be right back" so I can't even do anything there right now

**Comment by JC5** [comment]
Oh that's good, I was afraid it was gone for a moment.

I don't think the database container is linked in some way to the app container. Could it be enough to stop the new container and start the old one?

**Comment by 8bitgentleman** [comment]
Thankfully the original db is still there :) I've stopped the new db and started the original one but there's no way to tell if it worked because when I navigate to the port that I have firefly at all I get is this screen
<img width="1364" alt="screen shot 2017-11-18 at 1 50 32 pm" src="https://user-images.githubusercontent.com/4028391/32983753-dc18e0a8-cc67-11e7-97d1-5216f5b8b732.png">

The logs have a new error "php_network_getaddresses: getaddrinfo failed: Name or service not know"
[ff3-apache2handler-2017-11-18.log](https://github.com/firefly-iii/firefly-iii/files/1484929/ff3-apache2handler-2017-11-18.log)

**Comment by JC5** [comment]
Open your app container and check what it says in the .env file under "DB_HOST".

Could it be that this is a reference to the wrong container? Different IP, different name?

**Comment by 8bitgentleman** [comment]
DB_HOST=firefly-db and DB_PORT=3306 (which is the same port that the firefly-db container has open)

Is that what it should be?

**Comment by JC5** [comment]
Perhaps it should be `firefly-db_ORIG` or something?

**Comment by alazare619** [comment]
Yea you can change the database link of the docker-compose to point to firefly-db_ORIG im not sure what you are using to manage docker as thats not stock docker...It could be something like maybe docker swarm or kuberentes or portainer or something which could be causing interference...not sure but the docker-compose I linked you should of worked just fine on stock docker...

**Comment by alazare619** [comment]
DB container is the SQL container and i'm guessing DB ORIG is the orginal one in the docker-compose file you would want to remove the entire section called docker-db and under the link subsection of firefly-app change that to firefly-db_orig or whatever the current name is

**Comment by 8bitgentleman** [comment]
I run firefly iii on a Synology NAS therefor I run the official Synology Docker build. As far as I know It's nothing terribly different from stock docker just a GUI that uses the same commands and shows the currently running containers. Firefly iii was running perfectly before I tried to update it so this shouldn't be an issue with docker -_-

You're correct DB ORIG is my original database container. I renamed it before I started trying to update just to be sure that it wouldn't be overwritten. Is linking the DB ORIG to the new APP container supposed to fix the "Be Right Back" error? After linking the two the "Be Right Back" screen doesn't change

**Comment by JC5** [comment]
Mmmm. Does the error change in the log files, perhaps?

**Comment by alazare619** [comment]
That would be why it created a new container because you changed the db it
was expecting.....you essentially just caused yourself more work then
necessary...so yes because the db it created is not seeded or anything its
throwing the be right back error message...in the docker config remove the
section regarding mysql:8 the entire section and just change the link to
the new db name.

**Comment by 8bitgentleman** [comment]
@alazare619 To clarify what lines in your docker config should I change? Just remove [line 4 with image: mysql:8](https://gist.github.com/alazare619/1ae36356322c7fda7a80f85542d066b9#file-gistfile1-txt-L4) and then change [line 25](https://gist.github.com/alazare619/1ae36356322c7fda7a80f85542d066b9#file-gistfile1-txt-L25) to reflect the name of my original db container? Is that right?

**Comment by alazare619** [comment]
https://gist.github.com/alazare619/1ae36356322c7fda7a80f85542d066b9

That should link it.

**Comment by 8bitgentleman** [comment]
I ran the new docker-compose.yml file but it returned an error saying 

> ERROR: Service "firefly-app" has a link to service "fireflyiii_firefly-db_1_ORIG" which does not exist.

I double checked to make sure that the original db was running with docker ps -a and it says that the container is in fact running

`Container ID | Image | Command | Created | Status | Ports | Names`

`8578e9d0525c | mysql:8 | "docker-entrypoint.sh" | 4 months ago | Up 4 minutes | 3306/tcp | fireflyiii_firefly-db_1_ORIG`

Is there a service inside the db container that needs to be started?

**Comment by alazare619** [comment]
fireflyiii_firefly-db_1_ORIG is not equal to fireflyiiimaster_firefly-db_1_ORIG

**Comment by 8bitgentleman** [comment]
Whoops there was a typo in my post, the error and the running container name are the same

**Comment by alazare619** [comment]
that would mean then that infact your fireflyiiimaster_firefly-db_1_ORIG was not created with docker-compose which opens a whole new bag of worms...

**Comment by alazare619** [comment]
So in a nutshell heres how docker compose works it expects to create,start,stop, all containers. If something takes over and 1 container depends on another container that container must be running with docker-compose...so whatever your gui is doing that would mean that infact your container and volume for fireflyiiimaster_firefly-db_1_ORIG is being managed by a diffrent docker init system...and docker-compose can not communicate with it because of this..I would point you to the docker-compose docks at this point.

**Comment by alazare619** [comment]
TL;DR because of the fact that you renamed the docker container it created a new database, old database is now orphaned, old database needs to be managed by docker-compose. I'm not sure at this point how to even continue clone the container then work with it I suppose this is getting into areas that are platform specific tho. Since you are managing this via synology I cant in good faith say what to do. Synology may infact have its own docker init system. I know other nas's like unraid and rockstor and freenas do as well...

**Comment by alazare619** [comment]
I suppose you could install portainer and use it to clone the sql container at this point https://www.portainer.io/install.html

Afterwords you can then attempt to use docker compose to start/stop the sql container atleast then you would have a backup...the chaning of the name really threw a wrench into everything....

**Comment by 8bitgentleman** [comment]
I'm running all these commands via the command line though, I'm not using the GUI at all. I guess if when the original db was created it used another method that could be causing the issue. In trying to be careful and backup my data I really screwed things over... I'll have to take another stab at this after Thanksgiving.

**Comment by JC5** [comment]
Another possibility is to export the database from the image and import it into a new one.

**Comment by 8bitgentleman** [comment]
@JC5 how would I do that?

**Comment by JC5** [comment]
Run `docker container ls` to get a list of containers, and then `docker exec -it <container> bash` to get a bash prompt into the correct container.

In the container, you can run the `mysqldump` command with the parameters from your .env file. I shall look for an example.

Not sure yet how to get the content of the dump out of the container but worst case you could copy and paste it.

**Comment by JC5** [comment]
MySQL dump command:

`mysqldump firefly_db -ufirefly_db -pfirefly_db_secret`

Which is based on the `.env` file from the other container.

You could use termbin.com to export your data, but it would be public: http://termbin.com/

If you want to use this, run `apt-get update && apt-get install netcat` in the mysql container

Then run `mysqldump firefly_db -ufirefly_db -pfirefly_db_secret | nc termbin.com 9999`

The result would be a (public!) dump of your DB such as this one: http://termbin.com/bvb4

**Comment by 8bitgentleman** [comment]
Instead of using termbin I dumped the command output into a file using
`mysqldump firefly_db -ufirefly_db -pfirefly_db_secret > dump.sql`

I then, from outside the container, copied the dumped sql file for safe keeping. 
`docker cp <containerId>:/file/path/within/container /host/path/target`

How would I go about importing this dump into a new image?

**Comment by JC5** [comment]
Isn't that the same process in reverse? Copy it into the container, then use 

`mysql -ufirefly_db -pfirefly_db_secret < file.sql`

Should work as long as the import file starts with `USE DATABASE firefly_db`

**Comment by 8bitgentleman** [comment]
I'm not very familiar with mysql so I wanted to make sure. My dump file is very long but it doesn't start with `USE DATABASE firefly_db`, it doesn't even contain that phrase anywhere in it. When I run `mysql -ufirefly_db -pfirefly_db_secret < file.sql` (with the correct file name)
it returns an error `ERROR 1046 (3D000) at line 22: No database selected`. Would it be helpful for you to see my mysqldump file?

**Comment by JC5** [comment]
Use this command:

```
mysql -ufirefly_db -pfirefly_db_secret --database=firefly_db < file.sql
```

**Comment by 8bitgentleman** [comment]
@JC5 I took another swing at this today and got considerably farther than before. Starting up a fresh new container with firefly iii I was able to import the old data using the mysql commands. Logging into firefly works fine and all my balances are as expected. The issue I'm running into is looking at my old transactions. If I go into an asset account and try to " Show all transactions without grouping them by date." I get the "Be Right Back" screen. Adding new transactions works fine the error only happens when I try to look at old ones. 

[Here is my log file. ](https://gist.github.com/8bitgentleman/246ed98e61de723201286dc3c97c0bb9) 

Also I don't say it before but thanks to both @JC5  and @alazare619 for helping me through this unexpectedly complicated process!

**Comment by JC5** [comment]
Permission denied error on the storage thing:

```
 file_put_contents(/var/www/firefly-iii/storage/ ... 
````

Surf to `/flush` on your Docker machine, does that help?

**Comment by 8bitgentleman** [comment]
Sorry what do you mean "surf to `/flush`" Is that a directory within the firefly container? What do I do there?

**Comment by JC5** [comment]
Sorry, it's an URL like so: localhost/flush

**Comment by 8bitgentleman** [comment]
That doesn't seem to have had any effect.
[Logs](https://gist.github.com/8bitgentleman/b1ff42dd99dbe9225a0e1f31ca799932)

**Comment by JC5** [comment]
Crap. Ok, try this.

List your containers with `docker container ls`, launch bash into it: `docker exec -it <container> bash` and run `rm -rf /var/www/firefly/storage/framework/cache/*`.

That should clear out the cache.

**Comment by 8bitgentleman** [comment]
Whoa that caused another major error. Now clicking into any asset account throws this screen
<img width="625" alt="screen shot 2017-12-10 at 12 13 32 pm" src="https://user-images.githubusercontent.com/4028391/33807362-a69cc5dc-dda3-11e7-937b-059a522d5fe4.png">
[logs](https://gist.github.com/8bitgentleman/c257a463ab8574f33179451ae842b034)

**Comment by JC5** [comment]
Oh come on! :D

It might be a permission thing but we really should not have to pick around a docker container like this. You could fire up a new instance again but let's try this last trick. SSH into your container thing again and:

```
mkdir /var/www/firefly-iii/storage/framework/cache/data
chown -R www-data:www-data /var/www && chmod -R 775 /var/www/firefly-iii/storage
```

These are commands from the install script so they might reset everything to working order.

**Comment by 8bitgentleman** [comment]
Wait I figured it out! There was a typo in the `rm -rf` command that I didn't catch. The actual command should be `rm -rf /var/www/firefly-iii/storage/framework/cache/*`

After catching that I reran the command and holy crap it's finally working! :)

**Comment by JC5** [comment]
Hi all,

I've updated the docker files and the instructions.

New Dockerfile has three volumes for uploads, export and database. Docker compose file is simplified and had a version upgrade. The instructions on the website will match the files and it should work more easily now. This includes upgrading your containers.

I can't preview the website yet but [this is the new Dockerfile](https://github.com/firefly-iii/firefly-iii/blob/develop/Dockerfile), this is [the new compose file](https://github.com/firefly-iii/firefly-iii/blob/develop/docker-compose.yml) and [this is the new entrypoint.sh](https://github.com/firefly-iii/firefly-iii/blob/develop/docker/entrypoint.sh).

I'm fairly sure all of you get a message because you auto-follow this thread. So I invite you to look at these files and give me feedback if necessary. 

:+1:

**Comment by alazare619** [comment]
Looks good.

**Comment by 8bitgentleman** [comment]
Do I need to do something special to update my docker instance to this new dockerfile version or should I still follow the [update instructions on the website](https://firefly-iii.org/using-docker.html)?

**Comment by JC5** [comment]
To be safe I would backup the database as before and create new instances and storage volumes. Then restore the database. To check if it works you could remove the containers and recreate them again. The data should still be in the volume(s).

**Comment by JC5** [comment]
Closed to clean up the list.
